032-projects/proposal.md

```diff
@@ -0,0 +1,274 @@
+
+```yaml
+name: booklit
+plan:
+- task: test
+```
+
+Assuming the `test` task looks like this:
+
+```yaml
+platform: linux
+
+image_resource: # ...
+
+inputs:
+- name: booklit
+
+run:
+  path: booklit/ci/tasks/scripts/test
+```
+
+This would work just fine - the task would receive the `booklit` input via the project resource.
+
+But what does this look like once I start adopting pipelines? Now I'll want to use `passed` constraints, so I need to define a resource that I can reference with `get` steps. My repo is called `booklit`, so I'll just call it `booklit` too, and create `ci/resources/booklit.yml`:
+
+```yaml
+type: git
+source:
+  uri: https://github.com/vito/booklit
+  branch: master
+```
+
+Now I have a *project* named `booklit` and a *resource* named `booklit`. This is where the weirdness begins, but it's not just a simple naming mistake - we'll get back to this point soon.
+
+So, next I add my pipeline, `ci/pipelines/booklit.yml`:
+
+```yaml
+jobs:
+- name: test
+  plan:
+  - get: booklit
+    trigger: true
+  - task: test
+
+- name: ship
+  plan:
+  - get: booklit
+    passed: [test]
+  # ...
+```
+
+And now we have a subtle problem. Both of these builds have a `get: booklit`, and that's going to override the project-provided artifact. Structurally, this is fine, since they're both the same repo, but there's one gotcha: these jobs will be using the task config and/or scripts provided by the `get` step, not from the project config. Now it's possible for the job's config to fall out of sync with the tasks that it's running, because the `passed` constraint will at times provide an older version to the `ship` job.
+
+This can cause errors - for example, if a job has been re-configured to have a different set of inputs than what the older task expects, or if a new task config is being used for an outdated task script. (Depending on how we implement project tasks. Maybe the config comes from the artifact at runtime, too.)
+
+Even worse - you could potentially make a mistake and have your project tasks always running against the code from the *project* resource, and not what's actually under test in the pipeline.
+
+So, what's the solution here? Really, it comes down to this: your product code and your CI config are two different things. This becomes clear right from when you had to write an explicit resource for it in order to use `passed` constraints. Applying `passed` constraints to the project itself doesn't really make much sense - the job config had to come from somewhere in the first place to learn about the `passed` constraints. It's a bit of a chicken-egg problem.
+
+Instead, these should be treated as two separate inputs. Your product code is an input alongside your CI code. This also lets you split the CI code out into a separate repo if you want, and use it for every branch of your product, rather than having a copy on each branch and risk having them fall out of sync.
```

Comment:
Unsolicited thought: presumably this allows separating CI from product code && using it for each branch, but also for using the same CI code for _multiple products_ with _multiple branches_ (e.g., using single config repo to handle  CI for like... 13 other git repos, each containing a microservice composing the entire software solution).